File: python/searchable-client/src/searchable_client/facets.py
"""Facet shard fetching, filter interpretation, and per-field doc-id
narrowing for the search client (mirrors packages/client/src/facets.ts).
"""

from typing import Any

from searchable_client.fetch import ShardCache, resolve_url
from searchable_client.types import FacetShard, Manifest, facet_shard_from_dict


def _is_range_filter(value: Any) -> bool:
    return isinstance(value, dict) and ("min" in value or "max" in value)


def _values_for(filters: dict[str, Any] | None, field_name: str) -> list[str]:
    raw = (filters or {}).get(field_name)
    if raw is None or _is_range_filter(raw):
        return []
    return raw if isinstance(raw, list) else [raw]


def _range_filter_for(filters: dict[str, Any] | None, field_name: str) -> dict[str, Any] | None:
    raw = (filters or {}).get(field_name)
    return raw if _is_range_filter(raw) else None


def _fetch_facet_shards(
    manifest: Manifest, cache: ShardCache, base_url: str, fields: list[str]
) -> dict[str, FacetShard]:
    result = {}
    for entry in manifest.shards_facets:
        if entry.field in fields:
            raw = cache.fetch_json(resolve_url(base_url, entry.file))
            result[entry.field] = facet_shard_from_dict(raw)
    return result


def _union_docs_for_field(
    facet_shards_by_field: dict[str, FacetShard], filters: dict[str, Any] | None, field_name: str
) -> set[int]:
    shard = facet_shards_by_field.get(field_name)
    if shard is None:
        return set()
    ids: set[int] = set()
    if shard.type == "range":
        range_filter = _range_filter_for(filters, field_name)
        if not range_filter:
            return ids
        for range_entry in shard.sorted or []:
            if (
                "min" in range_filter
                and range_filter["min"] is not None
                and range_entry.value < range_filter["min"]
            ):
                continue
            if (
                "max" in range_filter
                and range_filter["max"] is not None
                and range_entry.value > range_filter["max"]
            ):
                continue
            ids.add(range_entry.doc)
        return ids
    for value in _values_for(filters, field_name):
        value_entry = shard.values.get(value)
        if value_entry:
            ids.update(value_entry.docs)
    return ids

File: packages/searchable/src/facets.ts
import type { ShardCache } from "./fetch-json.js";
import type { FacetShard, Manifest } from "./format/index.js";
import type { RangeFilter } from "./search.js";
import { resolve } from "./url.js";

function isRangeFilter(value: unknown): value is RangeFilter {
  return (
    typeof value === "object" &&
    value !== null &&
    !Array.isArray(value) &&
    ("min" in value || "max" in value)
  );
}

export function valuesFor(
  filters: Record<string, string | string[] | RangeFilter> | undefined,
  field: string,
): string[] {
  const raw = filters?.[field];
  if (raw === undefined || isRangeFilter(raw)) return [];
  return Array.isArray(raw) ? raw : [raw];
}

function rangeFilterFor(
  filters: Record<string, string | string[] | RangeFilter> | undefined,
  field: string,
): RangeFilter | undefined {
  const raw = filters?.[field];
  return isRangeFilter(raw) ? raw : undefined;
}

/** Fetches every facet shard in `fields` that actually exists in the manifest, keyed by field name. Shared by search() and facetValues() so they resolve facet shards identically. */
export async function fetchFacetShards(
  manifest: Manifest,
  cache: ShardCache,
  baseUrl: string,
  fields: string[],
): Promise<Map<string, FacetShard>> {
  const entries = (manifest.shards.facets ?? []).filter((s) =>
    fields.includes(s.field),
  );
  const fetched = await Promise.all(
    entries.map(
      async (entry) =>
        [
          entry.field,
          await cache.fetchJson<FacetShard>(resolve(baseUrl, entry.file)),
        ] as const,
    ),
  );
  return new Map(fetched);
}

/**
 * Every doc id matching the active filter on `field` (terms: OR across
 * selected values; range: min/max scan of the sorted array). Shared by
 * search() (candidate narrowing) and facetValues() (contextual counts
 * against every *other* active filter).
 */
export function unionDocsForField(
  facetShardsByField: Map<string, FacetShard>,
  filters: Record<string, string | string[] | RangeFilter> | undefined,
  field: string,
): Set<number> {
  const shard = facetShardsByField.get(field);
  const ids = new Set<number>();
  if (!shard) return ids;
  if (shard.type === "range") {
    const range = rangeFilterFor(filters, field);
    if (!range) return ids;
    // Linear scan over the sorted array rather than a binary-search
    // range lookup -- correct either way since the array is sorted,
    // and a full scan is negligible at "small corpus" JSON-tier
    // scale (docs/guides/indexing.md#what-to-simplify-at-this-scale).
    // Binary-searching the two endpoints is a documented future
    // optimization once shard size actually makes it matter.
    for (const entry of shard.sorted ?? []) {
      if (range.min !== undefined && entry.value < range.min) continue;
      if (range.max !== undefined && entry.value > range.max) continue;
      ids.add(entry.doc);
    }
    return ids;
  }
  for (const value of valuesFor(filters, field)) {
    for (const id of shard.values[value]?.docs ?? []) ids.add(id);
  }
  return ids;
}
